publisher/publisher.go

```diff
@@ -6,273 +6,136 @@
-	"bytes"
-	"crypto/ecdsa"
-	"crypto/sha256"
 	"crypto/x509"
 	"encoding/base64"
-	"encoding/json"
 	"fmt"
-	"io/ioutil"
 	"net/http"
-	"strconv"
 	"strings"
-	"time"
+
+	ct "github.com/letsencrypt/boulder/Godeps/_workspace/src/github.com/google/certificate-transparency/go"
+	ctClient "github.com/letsencrypt/boulder/Godeps/_workspace/src/github.com/google/certificate-transparency/go/client"
 
 	"github.com/letsencrypt/boulder/core"
 	blog "github.com/letsencrypt/boulder/log"
-	"github.com/letsencrypt/boulder/sa"
 )
 
-// LogDescription tells you how to connect to a log and verify its statements.
-type LogDescription struct {
-	ID        string
-	URI       string
-	PublicKey *ecdsa.PublicKey
-}
-
-type rawLogDescription struct {
-	URI       string `json:"uri"`
-	PublicKey string `json:"key"`
+// Log contains the CT client and signature verifier for a particular CT log
+type Log struct {
+	Client   *ctClient.LogClient
+	Verifier *ct.SignatureVerifier
 }
 
-// UnmarshalJSON parses a simple JSON format for log descriptions.  Both the
-// URI and the public key are expected to be strings.  The public key is a
-// base64-encoded PKIX public key structure.
-func (logDesc *LogDescription) UnmarshalJSON(data []byte) error {
-	var rawLogDesc rawLogDescription
-	if err := json.Unmarshal(data, &rawLogDesc); err != nil {
-		return fmt.Errorf("Failed to unmarshal log description, %s", err)
+// NewLog returns a initialized Log struct
+func NewLog(uri, b64PK string) (*Log, error) {
+	var l Log
+	var err error
+	if strings.HasSuffix(uri, "/") {
+		uri = uri[0 : len(uri)-2]
 	}
-	logDesc.URI = rawLogDesc.URI
-	// Load Key
-	pkBytes, err := base64.StdEncoding.DecodeString(rawLogDesc.PublicKey)
+	l.Client = ctClient.New(uri)
+
+	pkBytes, err := base64.StdEncoding.DecodeString(b64PK)
 	if err != nil {
-		return fmt.Errorf("Failed to decode base64 log public key")
+		return nil, fmt.Errorf("Failed to decode base64 log public key")
 	}
 	pk, err := x509.ParsePKIXPublicKey(pkBytes)
 	if err != nil {
-		return fmt.Errorf("Failed to parse log public key")
-	}
-	ecdsaKey, ok := pk.(*ecdsa.PublicKey)
-	if !ok {
-		return fmt.Errorf("Failed to unmarshal log description for %s, unsupported public key type", logDesc.URI)
+		return nil, fmt.Errorf("Failed to parse log public key")
 	}
-	logDesc.PublicKey = ecdsaKey
 
-	// Generate key hash for log ID
-	pkHash := sha256.Sum256(pkBytes)
-	logDesc.ID = base64.StdEncoding.EncodeToString(pkHash[:])
-	if len(logDesc.ID) != 44 {
-		return fmt.Errorf("Invalid log ID length [%d]", len(logDesc.ID))
-	}
-
-	return nil
-}
-
-// CTConfig defines the JSON configuration file schema
-type CTConfig struct {
-	Logs                       []LogDescription `json:"logs"`
-	SubmissionRetries          int              `json:"submissionRetries"`
-	SubmissionBackoffString    string           `json:"submissionBackoff"`
-	IntermediateBundleFilename string           `json:"intermediateBundleFilename"`
+	l.Verifier, err = ct.NewSignatureVerifier(pk)
```

Comment:
please combine all the sets here into just an allocation of Log{}. The l var is very far away with lots of stuff in there.